src/util/stats_base.h

```diff
@@ -0,0 +1,282 @@
+ * outside the statistics package (for example, one that's kept
+ * by a theory implementation for internal heuristic purposes,
+ * which is important to keep even if statistics are turned off).
+ *
+ * Template class T must have an assignment operator=().
+ */
+template <class T>
+class ReferenceStat : public Stat
+{
+ private:
+  /** The referenced data cell */
+  const T* d_data = nullptr;
+
+ public:
+  /**
+   * Construct a reference stat with the given name and a reference
+   * to nullptr.
+   */
+  ReferenceStat(const std::string& name) : Stat(name) {}
+
+  /**
+   * Construct a reference stat with the given name and a reference to
+   * the given data.
+   */
+  ReferenceStat(const std::string& name, const T& data) : Stat(name)
+  {
+    set(data);
+  }
+
+  /** Set this reference statistic to refer to the given data cell. */
+  void set(const T& t)
+  {
+    if (CVC4_USE_STATISTICS)
+    {
+      d_data = &t;
+    }
+  }
+  const T& get() const { return *d_data; }
+
+  /** Flush the value of the statistic to the given output stream. */
+  virtual void flushInformation(std::ostream& out) const override
+  {
+    out << *d_data;
+  }
+
+  virtual void safeFlushInformation(int fd) const override
+  {
+    safe_print<T>(fd, *d_data);
+  }
+
+}; /* class ReferenceStat<T> */
+
+/**
+ * A backed integer-valued (64-bit signed) statistic.
+ * This doesn't functionally differ from its base class BackedStat<int64_t>,
+ * except for adding convenience functions for dealing with integers.
+ */
+class IntStat : public BackedStat<int64_t>
+{
+ public:
+  /**
+   * Construct an integer-valued statistic with the given name and
+   * initial value.
+   */
+  IntStat(const std::string& name, std::int64_t init);
+
+  /** Increment the underlying integer statistic. */
+  IntStat& operator++();
+  /** Increment the underlying integer statistic. */
+  IntStat& operator++(int);
+
+  /** Increment the underlying integer statistic by the given amount. */
+  IntStat& operator+=(std::int64_t val);
+
+  /** Keep the maximum of the current statistic value and the given one. */
+  void maxAssign(int64_t val);
+
+  /** Keep the minimum of the current statistic value and the given one. */
+  void minAssign(int64_t val);
+
+  SExpr getValue() const override { return SExpr(Integer(d_data)); }
+
+}; /* class IntStat */
+
+/**
+ * The value for an AverageStat is the running average of (e1, e_2, ..., e_n),
+ *   (e1 + e_2 + ... + e_n)/n,
+ * where e_i is an entry added by an addEntry(e_i) call.
+ * The value is initially always 0.
+ * (This is to avoid making parsers confused.)
+ *
+ * A call to setData() will change the running average but not reset the
+ * running count, so should generally be avoided.  Call addEntry() to add
+ * an entry to the average calculation.
+ */
+class AverageStat : public BackedStat<double>
+{
+ private:
+  /**
+   * The number of accumulations of the running average that we
+   * have seen so far.
+   */
+  std::uint32_t d_count = 0;
```

Comment:
Iirc we agreed in a discussion offline that we don't prefix integer types with `std::`.